### PR TITLE
4396 long polling blocked bug - target 0.23.0

### LINKED ISSUE #4396: Long Polling is blocked even though jobs are available on the broker
**Describe the bug**

While using the java client to complete jobs we observed that sometimes the job worker will hang for the long polling timeout although jobs are available in the broker.

**To Reproduce**

Not totally clear. I often saw it by having a constant stream of create workflow instances and then at the same time a job worker running. Periodically the job worker was waiting for up to 20s before failing to activate jobs and then on repolling it found new jobs.

**Expected behavior**

Job workers should always be able to acquire jobs if they are available without running in any blocking timeout.

**Environment:**
 - OS: Linux/Camunda Cloud
 - Zeebe Version: 0.23.0
 - Configuration: default

### PR DESCRIPTION
**Pull request by pihme**
## Description

This PR proposes a fix to the long polling blocked bug. I don't expect it be approved right away. Instead I hope for some critical comments and engaging discussion.

Please read the issue and the analysis therein to understand what was fixed.

Talking about the fixes:
- for the backpressure the `ActivateJobsHandler` now tracks whether some brokers returned a resource exhausted error and if so, it tells the polling handler about it. I think this change is solid, albeit not very elegant.
- for the time window where the long polling handler did not know that a request is currently running: Here the main idea is to register any active request right away. I think the approach is solid, but the implementation is somewhat lacking. It is quite big for a simple bugfix, and rather small for a major refactoring. Also, besides the intended change in behavior there are some unintended side effects. Like scheduling of long polling requests will change slightly in order. And the number of how often the Broker is queried also changes. The existing code had some downsides, in my opinion, so I took some liberty in deviating from the existing implementation. Having said that, it is hard to know which specific behavior was intended and which was arbitrary. 

Happy to discuss the details and answer any questions. Just want to create the review early given that it is a critical bug and I did the best I could.

## Related issues

closes #4396

## Pull Request Checklist

- [X] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [X] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [X] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Comment by menski** [comment]
Looping in @deepthidevaki as she was the main author of the original feature. Maybe she has some insides to share.

File: gateway/src/main/java/io/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java [hunk 1]
```diff
@@ -77,87 +80,110 @@ public void activateJobs(
...
+      } else {
+        actor.submit(
+            () -> {
+              state.incrementFailedAttempts(currentTimeMillis());
+              activateJobs(request);
+            });
+      }
     } else {
       actor.submit(request::complete);
+      actor.submit(() -> handlePendingRequests(request.getType()));
```
Comment (Zelldon) [inline gateway/src/main/java/io/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java:154]: So we just continue with sending requests?

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Thanks @pihme I just had a look code wise and found some issues we can discuss.

Do you think it makes sense to maybe refactor a bit more the current LongPollingHandler, because it was a bit hard to understand to review it. Or should we do it in a separate issue?

Cool that you found a way to test it. Do you think we can somehow simplify the test ? Does this test fail without your changes?

How you fixed the issue with the missed notification during checking the brokers? Is this because you just retry on onComplete?

File: gateway/src/main/java/io/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java [hunk 2]
```diff
@@ -77,87 +80,110 @@ public void activateJobs(
...
-      unblockRequests(removedState);
-    }
-  }
+    state.resetFailedAttempts(0);
 
-  private void unblockRequests(final JobTypeAvailabilityState state) {
-    final Queue<LongPollingActivateJobsRequest> requests = state.getBlockedRequests();
-    if (requests == null) {
-      return;
+    if (state.getActiveRequest() == null) {
```
Comment (deepthidevaki) [inline gateway/src/main/java/io/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java:169]: Isn't this again preventing concurrent requests?

**Review by deepthidevaki — COMMENTED** [review]
I like the solution for handling backpressure. For the second case, I think the idea is good. However the current implementation prevents concurrent activate job requests to be send to the brokers. This will impact the overall job activation latency in normal execution when there are plenty of jobs available.

File: gateway/src/main/java/io/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java [hunk 3]
```diff
@@ -77,87 +80,110 @@ public void activateJobs(
...
         () -> {
-          final JobTypeAvailabilityState state = jobTypeState.get(request.getType());
-          final boolean isJobAvailable =
-              state == null || (state.getEmptyResponses() < emptyResponseThreshold);
-          if (isJobAvailable) {
-            activateJobsUnchecked(request);
+          final InFlightLongPollingActivateJobsRequestsState state =
+              getJobTypeState(request.getType());
+
+          if (state.getActiveRequest() == null
```
[the remark this answers is not in the document]
Comment (pihme) [reply gateway/src/main/java/io/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java]: You are so right. I focused so much on the edge case that I forgot the main case.

**Comment by pihme** [comment]
## Baseline
grafik

## Change
grafik

Not sure what to make of the two benchmarks. They are in the same ballpark. Also note that the changed benchmark ran for longer, whereas I took a screenshot of the baseline right when it started to dip in throughput.

**Review by deepthidevaki — APPROVED** [review]
LGTM. It would be also interesting to run a benchmark with low load - create only a few instances per sec - and verify workflow processing latency.

**Comment by pihme** [comment]
@menski Do you want to participate in this reivew?

**Comment by menski** [comment]
@pihme I don't have time right now to focus on that so please don't wait for me

**Comment by pihme** [comment]
I ran several benchmarks. High throughput, low throughput, settings that were close to menskis setup when he observed the bug, settings that were tailored in an attempt to provoke the bug. 

All I can say is that the software still runs with the changes and performance is within general fluctuation.

I was not able to reproduce the bug or show any significant improvement by the change. I do believe that the benchmarks currently are too blunt a tool to make these precise observations.

File: gateway/src/main/java/io/zeebe/gateway/impl/job/ActivateJobsHandler.java [hunk 4]
```diff
@@ -50,9 +51,16 @@ private void activateJobs(
...
+      final BiConsumer<Integer, Boolean> onCompleted) {
     activateJobs(
-        request, partitionIdIterator, remainingAmount, jobType, onResponse, onCompleted, false);
+        request,
+        partitionIdIterator,
+        remainingAmount,
+        jobType,
+        onResponse,
+        onCompleted,
+        false,
```
[the remark this answers is not in the document]
Comment (pihme) [reply gateway/src/main/java/io/zeebe/gateway/impl/job/ActivateJobsHandler.java:62]: See #4824

**Comment by pihme** [comment]
@npepinpe I cannot access the details for the security/snyk tests. 

Could you have a look at it and let me know what I need to know or whether I can ignore it?

**Comment by npepinpe** [comment]
It complains about the log4j2 dependency having problems. If you rebase it should be fixed as develop has the updated version, but you can also ignore it.

**Comment by pihme** [comment]
@npepinpe I am on 0.23 not develop, should this be fixed there as well?

**Comment by npepinpe** [comment]
Oh right, yeah we should back port it. I can open an issue for it, or we can do it here? It's updating log4j2 to 2.13.3.

**Comment by pihme** [comment]
I can do it with this PR

**Review by ChrisKujawa — APPROVED** [review]
:+1:

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by npepinpe** [comment]
If you're sure you can merge it yourself, looks like you've hit a particularly bad patch of flaky tests :S

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
bors retry

**Comment by pihme** [comment]
Merged manually due to unstable 0.23 branch

**Pull request merged**